src/Parser/Core/Modules/SpellUsable.js

```diff
@@ -0,0 +1,247 @@
+      switch (event.trigger) {
+        case 'begincooldown':
+          console.log(fightDuration, 'Cooldown started:', spellName(spellId), spellId, 'expected duration:', event.expectedDuration, `(charges on cooldown: ${this._currentCooldowns[spellId].chargesOnCooldown})`);
+          break;
+        case 'addcooldowncharge':
+          console.log(fightDuration, 'Used another charge:', spellName(spellId), spellId, `(charges on cooldown: ${this._currentCooldowns[spellId].chargesOnCooldown})`);
+          break;
+        case 'refreshcooldown':
+          console.log(fightDuration, 'Cooldown refreshed:', spellName(spellId), spellId);
+          break;
+        case 'restorecharge':
+          console.log(fightDuration, 'Charge restored:', spellName(spellId), spellId, `(charges left on cooldown: ${this._currentCooldowns[spellId].chargesOnCooldown})`);
+          break;
+        case 'endcooldown':
+          console.log(fightDuration, 'Cooldown finished:', spellName(spellId), spellId);
+          break;
+        default: break;
+      }
+    }
+
+    this.owner.triggerEvent(eventType, event);
+  }
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    this.beginCooldown(spellId, event.timestamp);
+  }
+  _checkCooldownExpiry(timestamp) {
+    Object.keys(this._currentCooldowns).forEach(spellId => {
+      const remainingDuration = this.cooldownRemaining(spellId, timestamp);
+      if (remainingDuration <= 0) {
+        const expectedEnd = this._currentCooldowns[spellId].start + this._currentCooldowns[spellId].expectedDuration;
+        const fightDuration = formatDuration((timestamp - this.owner.fight.start_time) / 1000);
+        debug && console.log(fightDuration, 'Clearing', spellName(spellId), spellId, 'due to expiry');
+        this.endCooldown(Number(spellId), false, expectedEnd);
+      }
+    });
+  }
+  _isCheckingCooldowns = false;
+  on_event(_, event) {
+    if (!this._isCheckingCooldowns) {
+      // This ensures this method isn't called again before it's finished executing (_checkCooldowns might trigger events).
+      this._isCheckingCooldowns = true;
+      this._checkCooldownExpiry((event && event.timestamp) || this.owner.currentTimestamp);
+      this._isCheckingCooldowns = false;
+    }
+  }
+  // Haste-based cooldowns gets longer/shorter when your Haste changes
+  // `newDuration = timePassed + (newCooldownDuration * (1 - progress))` (where `timePassed` is the time since the spell went on cooldown, `newCooldownDuration` is the full cooldown duration based on the new Haste, `progress` is the percentage of progress cooling down the spell)
+  on_changehaste(event) {
+    Object.keys(this._currentCooldowns).forEach(spellId => {
+      const cooldown = this._currentCooldowns[spellId];
+      const originalExpectedDuration = cooldown.expectedDuration;
+      const remainingDuration = this.cooldownRemaining(spellId, event.timestamp);
+      const timePassed = originalExpectedDuration - remainingDuration;
```

Comment:
`timePassed` can be simplified to `event.timestamp - cooldown.start`, which would remove the need to call `this.cooldownRemaining`.

Reply:
I prefer the way since it closely matches what I want and therefore is less likely to break. I'm pretty sure calling `cooldownRemaining` doesn't really cost anything noticeable, so maintainability is the major point here.